File: python/384.打乱数组.py
#
# @lc app=leetcode.cn id=384 lang=python3
#
# [384] 打乱数组
#


# @lc code=start
class Solution:

    # Fisher-Yates 洗牌算法
    def __init__(self, nums: List[int]):
        self.origin = nums.copy()
        self.nums = nums

    def reset(self) -> List[int]:
        self.nums = self.origin.copy()
        return self.nums

    def shuffle(self) -> List[int]:
        n = len(self.nums)
        for i in range(n):
            index = random.randrange(i, n)
            self.nums[index], self.nums[i] = self.nums[i], self.nums[index]
        return self.nums

'''
    # 暴力法
    def __init__(self, nums: List[int]):
        self.origin = nums.copy()
        self.nums = nums

    def reset(self) -> List[int]:
        self.nums = self.origin.copy()
        return self.nums

    def shuffle(self) -> List[int]:
        n = len(self.origin)
        shuffled = [0] * n
        for i in range(n):
            index = random.randrange(len(self.nums))
            shuffled[i] = self.nums.pop(index)
        self.nums = shuffled
        return self.nums

    # 调用api
    def __init__(self, nums: List[int]):
        self.ori = nums.copy()
        self.input = nums.copy()

    def reset(self) -> List[int]:
        return self.ori

    def shuffle(self) -> List[int]:
        random.shuffle(self.input)
        return self.input
'''
# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()
# @lc code=end

File: cpp/384.打乱数组.cpp
/*
 * @lc app=leetcode.cn id=384 lang=cpp
 *
 * [384] 打乱数组
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution {
public:
    vector<int> origin;
    vector<int> num;

    Solution(vector<int>& nums) {
        this->num = nums;
        this->origin.resize(nums.size());
        copy(nums.begin(), nums.end(), origin.begin());
    }
    
    vector<int> reset() {
        copy(origin.begin(), origin.end(), num.begin());
        return num;
    }
    
    vector<int> shuffle() {
        for (int i = 0; i < num.size(); i++) {
            int j = i + rand() % (num.size() - i);
            swap(num[i], num[j]);
        }
        return num;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */
// @lc code=end

